pkg/imageregistry/imageregistry.go

```diff
@@ -0,0 +1,233 @@
+		glog.V(100).Infof("The apiClient is empty")
+
+		return nil, fmt.Errorf("imageregistry Config 'apiClient' cannot be empty")
+	}
+
+	if imageRegistryObjName == "" {
+		glog.V(100).Infof("The name of the imageRegistry is empty")
+
+		return nil, fmt.Errorf("imageRegistry 'imageRegistryObjName' cannot be empty")
+	}
+
+	glog.V(100).Infof(
+		"Pulling imageRegistry object name: %s", imageRegistryObjName)
+
+	builder := Builder{
+		apiClient: apiClient,
+		Definition: &imageregistryv1.Config{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: imageRegistryObjName,
+			},
+		},
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("imageRegistry object %s doesn't exist", imageRegistryObjName)
+	}
+
+	builder.Definition = builder.Object
+
+	return &builder, nil
+}
+
+// Get fetches existing imageRegistry from cluster.
+func (builder *Builder) Get() (*imageregistryv1.Config, error) {
+	if valid, err := builder.validate(); !valid {
+		return nil, err
+	}
+
+	glog.V(100).Infof("Pulling existing imageRegistry with name %s from cluster", builder.Definition.Name)
+
+	imageRegistry := &imageregistryv1.Config{}
+	err := builder.apiClient.Get(context.TODO(), goclient.ObjectKey{
+		Name: builder.Definition.Name,
+	}, imageRegistry)
+
+	if err != nil {
+		glog.V(100).Infof("ImageRegistry object %s doesn't exist", builder.Definition.Name)
+
+		return nil, err
+	}
+
+	return imageRegistry, nil
+}
+
+// Exists checks whether the given imageRegistry exists.
+func (builder *Builder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if imageRegistry %s exists", builder.Definition.Name)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// Update renovates a imageRegistry in the cluster and stores the created object in struct.
+func (builder *Builder) Update() (*Builder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Updating the imageRegistry %s", builder.Definition.Name)
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("imageRegistry object %s doesn't exist", builder.Definition.Name)
+	}
+
+	err := builder.apiClient.Update(context.TODO(), builder.Definition)
+	if err == nil {
+		builder.Object = builder.Definition
+	}
+
+	return builder, err
+}
+
+// GetManagementState fetches imageRegistry ManagementState.
+func (builder *Builder) GetManagementState() (v1.ManagementState, error) {
+	if valid, err := builder.validate(); !valid {
+		return "", err
+	}
+
+	glog.V(100).Infof("Getting imageRegistry ManagementState configuration")
+
+	if !builder.Exists() {
+		return "", fmt.Errorf("imageRegistry object doesn't exist")
+	}
+
+	return builder.Object.Spec.ManagementState, nil
+}
+
+// WithManagementState returns true when succeeded to change imageRegistry operator management state.
+func (builder *Builder) WithManagementState(expectedManagementState v1.ManagementState) *Builder {
```

Comment:
Have you seen cases where the management state isn't updated correctly after using WithManagementState and Update? It seems like it would be more appropriate for this method to be WaitForManagementState that takes a timeout and desired management state and waits for it to be updated correctly. Then you could just call it after Update

Reply:
No, I personally have not encountered such a situation.
 The ManagementState was always immediately changed. Regarding the storage configuration, yes, it takes some time. But this method is outside of this package coverage, according to Nikita's request, since not only imagestream object is used but also others involved.

Reply:
Sorry I meant to add this comment to SetManagementState

Reply:
Could you change SetManagementState to something like WaitForManagementState? I think you just need a method that waits for the resource to get updated correctly.

Reply:
Whatever you say, I can do it, of course, but it's pointless. I already left a comment for Sasha: the "SetManagementState" method is an action, not a review, and it does not wait.
verify if ManagementState is already as expected -> exit with true
if not -> execute builder.WithManagementState(expectedManagementState).Update()
verify if ManagementState was changed (no wait needed)
In addition, we'll add a meanless parameter we'll need to take care of after.

Reply:
https://github.com/openshift-kni/eco-goinfra/pull/352#discussion_r1572940340

Reply:
moving this method to the eco-gotests repo